integrations/go/workload-executor.go

```diff
@@ -0,0 +1,315 @@
+		default:
+			return nil, errors.New("unrecognized updateOne option: " + key)
+		}
+	}
+	if opts.Upsert == nil {
+		opts = opts.SetUpsert(false)
+	}
+
+	return coll.UpdateOne(context.Background(), filter, update, opts)
+}
+
+func verifyInsertOneResult(actualResult *mongo.InsertOneResult, expectedResult interface{}) bool {
+	if expectedResult == nil {
+		return true
+	}
+
+	var expected mongo.InsertOneResult
+	if bson.Unmarshal(expectedResult.(bson.Raw), &expected) != nil {
+		return false
+	}
+
+	expectedID := expected.InsertedID
+	if f, ok := expectedID.(float64); ok && f == math.Floor(f) {
+		expectedID = int32(f)
+	}
+
+	return expectedID == nil || (actualResult != nil && expectedID == actualResult.InsertedID)
+}
+
+func verifyCursorResult(cur *mongo.Cursor, result interface{}) bool {
+	if result == nil {
+		return true
+	}
+
+	if cur == nil {
+		return false
+	}
+
+	defer func() {
+		cur.Close(context.Background())
+	}()
+
+	for _, expected := range result.(bson.A) {
+		if !cur.Next(context.Background()) {
+			return false
+		}
+		if !bytes.Equal(expected.(bson.Raw), cur.Current) {
+			return false
+		}
+	}
+
+	if cur.Next(context.Background()) {
+		return false
+	}
+	return cur.Err() == nil
+}
+
+func verifyUpdateResult(res *mongo.UpdateResult, result interface{}) bool {
+	if result == nil {
+		return true
+	}
+
+	var expected struct {
+		MatchedCount  int64 `bson:"matchedCount"`
+		ModifiedCount int64 `bson:"modifiedCount"`
+		UpsertedCount int64 `bson:"upsertedCount"`
+	}
+	err := bson.Unmarshal(result.(bson.Raw), &expected)
+	if err != nil {
+		return false
+	}
+
+	if expected.MatchedCount != res.MatchedCount {
+		return false
+	}
+	if expected.ModifiedCount != res.ModifiedCount {
+		return false
+	}
+
+	actualUpsertedCount := int64(0)
+	if res.UpsertedID != nil {
+		actualUpsertedCount = 1
+	}
+	return expected.UpsertedCount == actualUpsertedCount
+}
+
+func executeCollectionOperation(coll *mongo.Collection, op *operation) (bool, error) {
+	switch op.Name {
+	case "insertOne":
+		res, err := executeInsertOne(coll, op.Arguments)
+		return verifyInsertOneResult(res, op.Result), err
+	case "find":
+		cursor, err := executeFind(coll, op.Arguments)
+		return verifyCursorResult(cursor, op.Result), err
+	case "updateOne":
+		res, err := executeUpdateOne(coll, op.Arguments)
+		return verifyUpdateResult(res, op.Result), err
+	}
+	return false, errors.New("unrecognized collection operation: " + op.Name)
+}
+
+func runOperation(coll *mongo.Collection, op *operation) (bool, error) {
+	// execute the command on the given object
+	if op.Object == "collection" {
+		return executeCollectionOperation(coll, op)
+	}
+	return false, errors.New("unrecognized object: " + op.Name)
+}
+
+func main() {
+	connstring := os.Args[1]
+	workloadSpec := os.Args[2]
+
+	var workload driverWorkload
+
+	err := bson.UnmarshalExtJSONWithRegistry(specTestRegistry,[]byte(workloadSpec), false, &workload)
+	if err != nil {
+		str := "failed to unmarshal data: " + err.Error()
```

Comment:
nit: You usually don't need to add contexts to panics because the panic output will have a stack trace with the line number. If you feel that these are helpful, I'm fine leaving them in, but would prefer using `fmt.Sprintf` here rather than string concatenation to avoid any weird spacing issues if we make these messages more complex in the future.

Reply:
done